libWlzExtFF/WlzExtFF.3
'\" t
.\" ident MRC HGU $Id$
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" Project:    Woolz
.\" Title:      WlzExtFF.3
.\" Date:       March 1999
.\" Author:     Bill Hill
.\" Copyright:	1999 Medical Research Council, UK.
.\"		All rights reserved.
.\" Address:	MRC Human Genetics Unit,
.\"		Western General Hospital,
.\"		Edinburgh, EH4 2XU, UK.
.\" Purpose:    Functions for reading and writting Woolz objects to
.\"		and from external data formats.
.\" $Revision$
.\" Maintenance:Log changes below, with most recent at top of list.
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH "WlzExtFF" 3 "MRC HGU Woolz" "Woolz Library (External Data Formats)"
.SH NAME
WlzExtFF \- Woolz external data format functions.
.SH SYNOPSIS
.LP
.B #include <Wlz.h>
.br
.B #include <WlzExtFF.h>
.LP
typedef enum
.br
{
.br
.in +4m
WLZEFF_FORMAT_NONE,
.br
WLZEFF_FORMAT_BMP,
.br
WLZEFF_FORMAT_DEN,
.br
WLZEFF_FORMAT_ICS,
.br
WLZEFF_FORMAT_PNM,
.br
WLZEFF_FORMAT_PIC,
.br
WLZEFF_FORMAT_SLC,
.br
WLZEFF_FORMAT_VFF,
.br
WLZEFF_FORMAT_VTK,
.br
WLZEFF_FORMAT_WLZ,
.br
WLZEFF_FORMAT_TIFF,
.br
WLZEFF_FORMAT_COUNT
.in -4m
.br
} WlzEffFormat;
.LP
.BI "WlzEffFormat WlzEffStringExtToFormat(const char *" extStr );
.LP
.BI "WlzEffFormat WlzEffStringToFormat(const char *" fmtStr );
.LP
.BI "const char *WlzEffStringFromFormat(WlzEffFormat " fFmt ,
.in +4m
.br
.BI "const char **" dstExtStr );
.in -4m
.LP
.BI "WlzObject *WlzEffReadObj(FILE *" fP ,
.in +4m
.br
.BI "const char *" "fName" ,
.BI "WlzEffFormat " fFmt ,
.br
.BI "WlzErrorNum *" dstErr );
.in -4m
.BI "WlzErrorNum WlzEffWriteObj(FILE *" fP ,
.in +4m
.br
.BI "const char *" "fName" ,
.BI "WlzObject *" obj ,
.br
.BI "WlzEffFormat " fFmt );
.in -4m
.SH ARGUMENTS
.TS
tab(^);
lI l.
dstErr^Destination pointer for Woolz error number, may be NULL.
dstExtStr^Destination pointer for extension string.
extStr^Given file extension string.
fFmt^Given file format enumerated type.
fName^File name or base for file names.
fP^File stream.
fmtStr^Given file format string.
obj^Given woolz object.
.TE
.SH MT-LEVEL
.LP
Unsafe
.SH DESCRIPTION
.LP
The utility functions \fBWlzEffStringExtToFormat\fR,
\fBWlzEffStringToFormat\fR and \fBWlzEffStringFromFormat\fR
allow conversion between the file format enumerated types and
text strings.
.SS WlzEffStringExtToFormat
Takes as a parameter a file extension string and returns the
corresponding file format enumerated type.
The return value
WLZEFF_FORMAT_NONE is used when there is no format corresponding 
to the given string.
.SS WlzEffStringToFormat
Takes as a parameter a file format string and returns the corresponding file
format enumerated type. 
The return value
WLZEFF_FORMAT_NONE is used when there is no format corresponding 
to the given string.
.SS WlzEffStringFromFormat
Given a file format enumerated type,
WlzEffStringFromFormat
returns a string which is suitable for displaying
(eg \'BioRad Confocal\')
and sets a destination pointer for the
formats file extension.
.SS WlzEffReadObj
Reads a woolz object from the the given file names(s)
or opened file stream, where the file format of the
given a parameter is used.
The BMP, ICS and PNM file formats require a file name, for
all other formats either the file name or a file stream
pointer may be given.
.SS WlzEffWriteObj
Writes a woolz object to the the given file names(s)
or opened file stream, where the file format of the
given a parameter is used.
The BMP, ICS and PNM file formats require a file name, for
all other formats either the file name or a file stream
pointer may be given.
.SH "FILE FORMATS"
.LP
.TS
tab(^);
cb s s s
cb s s s
lb lb lb lb
l l l l.
File Formats Supported.

String^Enumerated type^Ext(s)^Multiple files
Microsoft Bitmap^WLZEFF_FORMAT_BMP^bmp^One 2D bmp file
^^^for each plane
^^^plus a single
^^^control file for
^^^3D.
Stanford Density^WLZEFF_FORMAT_DEN^den^No.
ICS^WLZEFF_FORMAT_ICS^ics^Pair of ics and
^^ids^ids data files.
PNM^WLZEFF_FORMAT_PNM^pgm^One 2D pgm file
^^^for each plane
^^^plus a single
^^^control file for
^^^3D.
^^pnm^
BioRad Confocal^WLZEFF_FORMAT_PIC^pic^No.
SLC^WLZEFF_FORMAT_SLC^slc^No.
Sunvision VFF^WLZEFF_FORMAT_VFF^vff^No.
Visualization Toolkit^WLZEFF_FORMAT_VTK^vtk^No.
VTK^^^
Woolz^WLZEFF_FORMAT_WLZ^wlz^No.
Tagged Image Format^WLZEFF_FORMAT_TIFF^tif^File name required,
^^^cannot be read from a
^^^file pointer. Only
^^^tested for a small
^^^selection of tif files.
.TE
.LP 
A control file is used to support 3D volumes using 2D data formats. This
is a plain ascii text file which specifies the volume's origin
and size, the voxel size and the 2D data files.
.LP 
The following example specifies how the eight listed pgm data files
are combined to form a 3D volume.
.LP
.ps -2
.cs R 24
.nf
ident:WLZSTACKCTR:1.0
# A comment
volume origin:10 30 50
# Volume starts at column 10, line 30, plane 50
volume size:120 120 8
# The volume has 120 columns, 150 lines and 8 planes
voxel size:1 1 1
files
       50 lob00000000.pgm
       51 lob00000001.pgm
       52 lob00000002.pgm
       53 lob00000003.pgm
       54 lob00000004.pgm
       55 lob00000005.pgm
       56 lob00000006.pgm
       57 lob00000007.pgm
.fi
.cs R
.ps +2
.LP
The first record identifies the file type and version,
if it is not correct the control file wil not be read.
.LP
The volume origin (default 0 0 0) and
voxel size (default 1.0 1.0 1.0) are optional but the
volume size \fBmust\fR be supplied.
Dimensions are given as <column> <line> <plane>.
Comments can be included using the '#' character.
.SH EXAMPLES
.LP
.ps -2
.cs R 24
.nf
/*
 * Simple example using WlzEffReadObj() and WlzEffWriteObj() to
 * convert Biorad '.pic' files into Sunvision '.vff' files.
 * The '.pic' file is read from the standard input and the '.vff'
 * file is written to the standard output.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <Wlz.h>
#include <WlzExtFF.h>

int             main(char *argv[], int argc)
{
  WlzObject     *obj = NULL;
  WlzErrorNum   errNum = WLZ_ERR_NONE;

  if(((obj = WlzEffReadObj(stdin, NULL, WLZEFF_FORMAT_PIC,
                           &errNum)) == NULL) || (errNum != WLZ_ERR_NONE))
  {
    (void )fprintf(stderr, "%s: Failed to read input object.\\n", *argv);
    errNum = EOF_READ_ERROR;
  }
  else if(WlzEffWriteObj(stdout, NULL, obj, WLZEFF_FORMAT_VFF) != WLZ_ERR_NONE)
  {
    (void )fprintf(stderr, "%s: Failed to write output object.\\n", *argv);
    errNum = EOF_WRITE_ERROR;
  }
  if(obj)
  {
    WlzFreeObj(obj);
  }
  return(errNum);
}
.fi
.cs R
.ps +2
.SH SEE ALSO
WlzError(3)
.SH BUGS
Still to be found.


ChangeLog
$Log$
Revision 1.2  2007/12/05 13:33:30  bill
*** empty log message ***



libWlzExtFF/WlzExtFF.c
#pragma ident "MRC HGU $Id$"
/***********************************************************************
* Project:      Woolz
* Title:        WlzExtFF.c
* Date:         March 1999
* Author:       Bill Hill
* Copyright:	1999 Medical Research Council, UK.
*		All rights reserved.
* Address:	MRC Human Genetics Unit,
*		Western General Hospital,
*		Edinburgh, EH4 2XU, UK.
* Purpose:      Functions for reading and writting Woolz objects to
*		and from external data formats.
* $Revision$
* Maintenance:	Log changes below, with most recent at top of list.
************************************************************************/
#include <Wlz.h>
#include <WlzExtFF.h>

/************************************************************************
* Function:	WlzEffStringExtToFormat					*
* Returns:	WlzEffFormat:		File format, WLZEFF_FORMAT_NONE	*
*					if string not matched.		*
* Purpose:	Returns a file format for the given file extension 	*
*		string.							*
* Global refs:	-							*
* Parameters:	const char *extStr:	Given file extension string.	*
************************************************************************/
WlzEffFormat	WlzEffStringExtToFormat(const char *extStr)
{
  unsigned int	fileFmt;

  if(WlzStringMatchValue((int *)&fileFmt, extStr,
			 "bmp", WLZEFF_FORMAT_BMP,
			 "den", WLZEFF_FORMAT_DEN,
			 "ics", WLZEFF_FORMAT_ICS,
			 "ids", WLZEFF_FORMAT_ICS,
			 "pgm", WLZEFF_FORMAT_PNM,
			 "pnm", WLZEFF_FORMAT_PNM,
			 "pic", WLZEFF_FORMAT_PIC,
			 "slc", WLZEFF_FORMAT_SLC,
			 "vff", WLZEFF_FORMAT_VFF,
			 "vtk", WLZEFF_FORMAT_VTK,
			 "wlz", WLZEFF_FORMAT_WLZ,
			 "ipl", WLZEFF_FORMAT_IPL,
			 "tif", WLZEFF_FORMAT_TIFF,
			 NULL) == 0)
  {
    fileFmt = (unsigned int )WLZEFF_FORMAT_NONE;
  }
  return((WlzEffFormat )fileFmt);
}

/************************************************************************
* Function:	WlzEffStringToFormat					*
* Returns:	WlzEffFormat:		File format, WLZEFF_FORMAT_NONE	*
*					if string not matched.		*
* Purpose:	Returns a file format for the given file format string.	*
* Global refs:	-							*
* Parameters:	const char *fmtStr:	Given file format string.	*
************************************************************************/
WlzEffFormat	WlzEffStringToFormat(const char *fmtStr)
{
  unsigned int	fileFmt;

  if(WlzStringMatchValue((int *)&fileFmt, fmtStr,
			 "Microsoft Bitmap", WLZEFF_FORMAT_BMP,
			 "Stanford Density", WLZEFF_FORMAT_DEN,
			 "ICS", WLZEFF_FORMAT_ICS,
			 "PNM", WLZEFF_FORMAT_PNM,
			 "BioRad Confocal", WLZEFF_FORMAT_PIC,
			 "SLC", WLZEFF_FORMAT_SLC,
			 "Sunvision VFF", WLZEFF_FORMAT_VFF,
			 "Visualization Toolkit VTK", WLZEFF_FORMAT_VTK,
			 "Woolz", WLZEFF_FORMAT_WLZ,
			 "IPLab", WLZEFF_FORMAT_IPL,
			 "Tiff", WLZEFF_FORMAT_TIFF,
			 NULL) == 0)
  {
    fileFmt = (unsigned int )WLZEFF_FORMAT_NONE;
  }
  return((WlzEffFormat )fileFmt);
}

/************************************************************************
* Function:	WlzEffStringFromFormat					*
* Returns:	const char *:		String appropriate to format,	*
*					or NULL on error.		*
* Purpose:	Returns a string which is suitable for displaying	*
*		the given format and sets a destination pointer for the	*
*		formats file extension.					*
* Global refs:	-							*
* Parameters:	WlzEffFormat fileFmt:	Given file format.		*
*		const char **dstExtStr:	Destination ptr for extension	*
*					string.				*
************************************************************************/
const char	*WlzEffStringFromFormat(WlzEffFormat fileFmt,
					const char **dstExtStr)
{
  const char	*extStr = NULL,
  		*fmtStr = NULL;
  const char	*extBmpStr = "bmp",
		*fmtBmpStr = "Microsoft Bitmap",
  		*extDenStr = "den",
  		*fmtDenStr = "Stanford Density",
  		*extIcsStr = "ics",
  		*fmtIcsStr = "ICS",
  		*extPnmStr = "pnm",
  		*fmtPnmStr = "PNM",
  		*extPicStr = "pic",
  		*fmtPicStr = "BioRad Confocal",
  		*extSlcStr = "slc",
		*fmtSlcStr = "SLC",
  		*extVffStr = "vff",
  		*fmtVffStr = "Sunvision VFF",
  		*extVtkStr = "vtk",
  		*fmtVtkStr = "Visualization Toolkit VTK",
		*extWlzStr = "wlz",
		*fmtWlzStr = "Woolz",
		*extIPLStr = "ipl",
		*fmtIPLStr = "IPLab",
		*extTiffStr = "tif",
		*fmtTiffStr = "Tiff";

  switch(fileFmt)
  {
    case WLZEFF_FORMAT_BMP:
      fmtStr = fmtBmpStr;
      extStr = extBmpStr;
      break;
    case WLZEFF_FORMAT_DEN:
      fmtStr = fmtDenStr;
      extStr = extDenStr;
      break;
    case WLZEFF_FORMAT_ICS:
      fmtStr = fmtIcsStr;
      extStr = extIcsStr;
      break;
    case WLZEFF_FORMAT_PNM:
      fmtStr = fmtPnmStr;
      extStr = extPnmStr;
      break;
    case WLZEFF_FORMAT_PIC:
      fmtStr = fmtPicStr;
      extStr = extPicStr;
      break;
    case WLZEFF_FORMAT_SLC:
      fmtStr = fmtSlcStr;
      extStr = extSlcStr;
      break;
    case WLZEFF_FORMAT_VFF:
      fmtStr = fmtVffStr;
      extStr = extVffStr;
      break;
    case WLZEFF_FORMAT_VTK:
      fmtStr = fmtVtkStr;
      extStr = extVtkStr;
      break;
    case WLZEFF_FORMAT_WLZ:
      fmtStr = fmtWlzStr;
      extStr = extWlzStr;
      break;
    case WLZEFF_FORMAT_IPL:
      fmtStr = fmtIPLStr;
      extStr = extIPLStr;
      break;
  }
  if(dstExtStr)
  {
    *dstExtStr = extStr;
  }
  return(fmtStr);
}

/************************************************************************
* Function:	WlzEffReadObj						*
* Returns:	WlzObject:		Woolz object read from file, or	*
*					NULL on error.			*
* Purpose:	Reads a woolz object from the the given file names(s)	*
*		or opened file stream, where the file format is that	*
*		given as a parameter.					*
*		The file ICS and PNM formats require a file name, for	*
*		all other formats either the file name or a file stream	*
*		pointer may be given.					*
* Global refs:	-							*
* Parameters:	FILE *fP:		Input file stream.		*
*		const char *fName:	File name (and path).		*
*		WlzEffFormat fFmt:	File format.			*
* 		WlzErrorNum *dstErr:	Destination error number ptr,	*
*					may be NULL.			*
************************************************************************/
WlzObject	*WlzEffReadObj(FILE *fP, const char *fName, WlzEffFormat fFmt,
			       WlzErrorNum *dstErr)
{
  int		openFileFlag = 0;
  WlzObject	*obj = NULL;
  WlzErrorNum	errNum = WLZ_ERR_NONE;

  if((fP == NULL) && (fName == NULL))
  {
    errNum = WLZ_ERR_PARAM_NULL;
  }
  else if(fP == NULL)
  {
    if((fFmt != WLZEFF_FORMAT_ICS) && (fFmt != WLZEFF_FORMAT_PNM) &&
       (fFmt != WLZEFF_FORMAT_BMP) && (fFmt != WLZEFF_FORMAT_TIFF))
    {
      if((fP = fopen(fName, "r")) == NULL)
      {
	errNum = WLZ_ERR_READ_EOF;
      }
      else
      {
	openFileFlag = 1;
      }
    }
  }
  if(errNum == WLZ_ERR_NONE)
  {
    switch(fFmt)
    {
      case WLZEFF_FORMAT_BMP:
	obj = WlzEffReadObjBmp(fName, &errNum);
	break;
      case WLZEFF_FORMAT_DEN:
	obj = WlzEffReadObjDen(fP, &errNum);
	break;
      case WLZEFF_FORMAT_ICS:
	obj = WlzEffReadObjIcs(fName, &errNum);
	break;
      case WLZEFF_FORMAT_PNM:
	obj = WlzEffReadObjPnm(fName, &errNum);
	break;
      case WLZEFF_FORMAT_PIC:
	obj = WlzEffReadObjPic(fP, &errNum);
	break;
      case WLZEFF_FORMAT_SLC:
	obj = WlzEffReadObjSlc(fP, &errNum);
	break;
      case WLZEFF_FORMAT_VFF:
	obj = WlzEffReadObjVff(fP, &errNum);
	break;
      case WLZEFF_FORMAT_VTK:
	obj = WlzEffReadObjVtk(fP, &errNum);
	break;
      case WLZEFF_FORMAT_WLZ:
        obj = WlzReadObj(fP, &errNum);
	break;
      case WLZEFF_FORMAT_IPL:
        obj = WlzEffReadObjIPL(fP, &errNum);
	break;
      case WLZEFF_FORMAT_TIFF:
	obj = WlzEffReadObjTiff(fName, &errNum);
	break;
      default:
        errNum = WLZ_ERR_PARAM_DATA;
	break;
    }
  }
  if(openFileFlag)
  {
    fclose(fP);
  }
  if(dstErr)
  {
    *dstErr = errNum;
  }
  return(obj);
}


/************************************************************************
* Function:	WlzEffWriteObj						*
* Returns:	WlzErrorNum:		Woolz error number.		*
* Purpose:	Writes a woolz object to the the given file stream,	*
*		with the given file format.				*
* Global refs:	-							*
* Parameters:	FILE *fP:		Output file stream.		*
*		const char *fName:	File name (and path).		*
*		WlzObject *obj:		Given woolz object.		*
*		WlzEffFormat fFmt:	File format.			*
************************************************************************/
WlzErrorNum	WlzEffWriteObj(FILE *fP, const char *fName, WlzObject *obj,
			       WlzEffFormat fFmt)
{
  int		openFileFlag = 0;
  WlzErrorNum	errNum = WLZ_ERR_NONE;

  if((fP == NULL) && (fName == NULL))
  {
    errNum = WLZ_ERR_PARAM_NULL;
  }
  else if(fP == NULL)
  {
    if((fFmt != WLZEFF_FORMAT_ICS) && (fFmt != WLZEFF_FORMAT_PNM) &&
       (fFmt != WLZEFF_FORMAT_BMP))
    {
      if((fP = fopen(fName, "w")) == NULL)
      {
	errNum = WLZ_ERR_WRITE_EOF;
      }
      else
      {
	openFileFlag = 1;
      }
    }
  }
  if(errNum == WLZ_ERR_NONE)
  {
    switch(fFmt)
    {
      case WLZEFF_FORMAT_BMP:
	errNum = WlzEffWriteObjBmp(fName, obj);
	break;
      case WLZEFF_FORMAT_DEN:
	errNum = WlzEffWriteObjDen(fP, obj);
	break;
      case WLZEFF_FORMAT_ICS:
	errNum = WlzEffWriteObjIcs(fName, obj);
	break;
      case WLZEFF_FORMAT_PNM:
	errNum = WlzEffWriteObjPnm(fName, obj);
	break;
      case WLZEFF_FORMAT_PIC:
	errNum = WlzEffWriteObjPic(fP, obj);
	break;
      case WLZEFF_FORMAT_SLC:
	errNum = WlzEffWriteObjSlc(fP, obj);
	break;
      case WLZEFF_FORMAT_VFF:
	errNum = WlzEffWriteObjVff(fP, obj);
	break;
      case WLZEFF_FORMAT_VTK:
	errNum = WlzEffWriteObjVtk(fP, obj);
	break;
      case WLZEFF_FORMAT_WLZ:
        errNum = WlzWriteObj(fP, obj);
	break;
      case WLZEFF_FORMAT_IPL:
	errNum = WlzEffWriteObjIPL(fP, obj);
	break;
      case WLZEFF_FORMAT_TIFF:
	errNum = WlzEffWriteObjTiff(fName, obj);
	break;
      default:
        errNum = WLZ_ERR_PARAM_DATA;
	break;
    }
  }
  if(openFileFlag)
  {
    fclose(fP);
  }
  return(errNum);
}
